File: hermes/triage.py
"""#19 — Hermes Credential Triage

Classifies a failure event into one of five categories and maps each
to a remediation strategy string the agent can act on.
"""
from __future__ import annotations

import logging
import re

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Failure categories and their remediation strategies
# ---------------------------------------------------------------------------

REMEDIATION: dict[str, str] = {
    "infrastructure": (
        "Rotate proxy or tunnel endpoint. Check VPN status and retry with "
        "a different proxy_region. Verify that the backend service is reachable."
    ),
    "site_change": (
        "DOM structure has likely changed. Inspect the latest Markdown dump "
        "and screenshot, then update Playwright selectors in engine.ts."
    ),
    "credential_invalid": (
        "Credential is permanently invalid (disabled/noaccount). Mark as "
        "tested-bad and dequeue. Do NOT retry."
    ),
    "rate_limited": (
        "Too many attempts detected. Back off for 5-10 minutes, rotate IP, "
        "and reduce concurrency before retrying."
    ),
    "unknown": (
        "Failure could not be categorised. Collect screenshots and logs, "
        "escalate to the full diagnostic pipeline."
    ),
}

# ---------------------------------------------------------------------------
# Pattern sets (case-insensitive)
# ---------------------------------------------------------------------------

_INFRA_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"proxy", re.I),
    re.compile(r"tunnel", re.I),
    re.compile(r"ECONNREFUSED", re.I),
    re.compile(r"ETIMEDOUT", re.I),
    re.compile(r"socket hang up", re.I),
    re.compile(r"network\s*error", re.I),
    re.compile(r"ERR_CONNECTION", re.I),
    re.compile(r"502|503|504", re.I),
]

_SITE_CHANGE_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"selector", re.I),
    re.compile(r"locator", re.I),
    re.compile(r"DOM", re.I),
    re.compile(r"element not found", re.I),
    re.compile(r"timeout waiting", re.I),
    re.compile(r"page\s*changed", re.I),
    re.compile(r"unexpected\s*(modal|dialog|popup)", re.I),
]

_CRED_INVALID_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"noaccount", re.I),
    re.compile(r"disabled", re.I),
    re.compile(r"invalid.*credential", re.I),
    re.compile(r"account.*not.*found", re.I),
    re.compile(r"deactivated", re.I),
]

_RATE_LIMIT_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"rate.?limit", re.I),
    re.compile(r"too many (attempts|requests)", re.I),
    re.compile(r"429", re.I),
    re.compile(r"throttl", re.I),
    re.compile(r"captcha", re.I),
]


def _matches(text: str, patterns: list[re.Pattern[str]]) -> bool:
    return any(p.search(text) for p in patterns)


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def classify_failure(event: dict) -> str:
    """Classify *event* into a failure category.

    Parameters
    ----------
    event:
        Raw websocket event dict (expects ``data.outcome`` and optionally
        ``data.error`` or ``data.message`` fields).

    Returns
    -------
    One of: ``'infrastructure'``, ``'site_change'``, ``'credential_invalid'``,
    ``'rate_limited'``, ``'unknown'``.
    """
    data = event.get("data", {})
    outcome = str(data.get("outcome", ""))
    error = str(data.get("error", ""))
    message = str(data.get("message", ""))
    blob = f"{outcome} {error} {message}"

    if _matches(blob, _CRED_INVALID_PATTERNS):
        category = "credential_invalid"
    elif _matches(blob, _RATE_LIMIT_PATTERNS):
        category = "rate_limited"
    elif _matches(blob, _SITE_CHANGE_PATTERNS):
        category = "site_change"
    elif _matches(blob, _INFRA_PATTERNS):
        category = "infrastructure"
    else:
        category = "unknown"

    logger.info("Triage classified event as '%s' — outcome='%s'", category, outcome)
    return category


def get_remediation(category: str) -> str:
    """Return the remediation strategy string for *category*."""
    return REMEDIATION.get(category, REMEDIATION["unknown"])

def cluster_errors(errors: list[str]) -> dict:
    """Group similar error strings together based on word-overlap (Jaccard similarity)."""
    clusters: dict[str, list[str]] = {}
    
    def get_words(text: str) -> set[str]:
        return set(re.findall(r'\w+', text.lower()))
        
    for error in errors:
        words = get_words(error)
        matched_cluster = None
        
        for cluster_rep in clusters:
            rep_words = get_words(cluster_rep)
            union_len = len(words | rep_words)
            if union_len == 0:
                if len(words) == 0:
                    matched_cluster = cluster_rep
                    break
                continue
            
            jaccard = len(words & rep_words) / union_len
            if jaccard > 0.5:
                matched_cluster = cluster_rep
                break
                
        if matched_cluster is not None:
            clusters[matched_cluster].append(error)
        else:
            clusters[error] = [error]
            
    return clusters

File: src/hermes/triage.ts
/**
 * #19 — Hermes Credential Triage (TypeScript Port)
 *
 * Classifies a failure event into one of five categories and maps each
 * to a remediation strategy string the agent can act on.
 *
 * Ported from hermes/triage.py
 */

// ---------------------------------------------------------------------------
// Failure categories and their remediation strategies
// ---------------------------------------------------------------------------

export type TriageCategory =
  | "infrastructure"
  | "site_change"
  | "credential_invalid"
  | "rate_limited"
  | "toxic_anomaly"
  | "unknown";

const REMEDIATION: Record<TriageCategory, string> = {
  infrastructure:
    "Rotate proxy or tunnel endpoint. Check VPN status and retry with " +
    "a different proxy_region. Verify that the backend service is reachable.",
  site_change:
    "DOM structure has likely changed. Inspect the latest Markdown dump " +
    "and screenshot, then update Playwright selectors in engine.ts.",
  credential_invalid:
    "Credential is permanently invalid (disabled/noaccount). Mark as " +
    "tested-bad and dequeue. Do NOT retry.",
  rate_limited:
    "Too many attempts detected. Back off for 5-10 minutes, rotate IP, " +
    "and reduce concurrency before retrying.",
  toxic_anomaly:
    "Toxic context detected (WAF block, 2FA, honeypot). Permanently poison context, " +
    "rotate proxy, and throttle worker to 1 active slot.",
  unknown:
    "Failure could not be categorised. Collect screenshots and logs, " +
    "escalate to the full diagnostic pipeline.",
};

// ---------------------------------------------------------------------------
// Pattern sets (case-insensitive)
// ---------------------------------------------------------------------------

const INFRA_PATTERNS: RegExp[] = [
  /proxy/i,
  /tunnel/i,
  /ECONNREFUSED/i,
  /ETIMEDOUT/i,
  /socket hang up/i,
  /network\s*error/i,
  /ERR_CONNECTION/i,
  /502|503|504/i,
  /crash/i,
];

const TOXIC_PATTERNS: RegExp[] = [
  /blocked/i,
  /403/i,
  /honeypot/i,
  /2fa/i,
  /mfa_required/i,
];

const SITE_CHANGE_PATTERNS: RegExp[] = [
  /selector/i,
  /locator/i,
  /DOM/i,
  /element not found/i,
  /timeout waiting/i,
  /page\s*changed/i,
  /unexpected\s*(modal|dialog|popup)/i,
];

const CRED_INVALID_PATTERNS: RegExp[] = [
  /noaccount/i,
  /permanently/i,
  /\bdisabled\b/i,
  /invalid.*credential/i,
  /account.*not.*found/i,
  /deactivated/i,
];

const RATE_LIMIT_PATTERNS: RegExp[] = [
  /rate.?limit/i,
  /too many (attempts|requests)/i,
  /429/i,
  /throttl/i,
  /captcha/i,
  /temporarily_disabled/i,
  /tempdisabled/i,
];

function matches(text: string, patterns: RegExp[]): boolean {
  return patterns.some((rx) => rx.test(text));
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export interface TriageEvent {
  data?: {
    outcome?: string;
    error?: string;
    message?: string;
    [key: string]: unknown;
  };
  [key: string]: unknown;
}

/**
 * Classify a failure event into a triage category.
 *
 * Returns one of: `infrastructure`, `site_change`, `credential_invalid`,
 * `rate_limited`, `unknown`.
 */
export function classifyFailure(event: TriageEvent): TriageCategory {
  const data = event.data ?? {};
  const outcome = String(data.outcome ?? "");
  const error = String(data.error ?? "");
  const message = String(data.message ?? "");
  const blob = `${outcome} ${error} ${message}`;

  // Rate limiting is checked before toxic to avoid false positives
  // (e.g., outcome="blocked" with error="rate limit" should be rate_limited, not toxic)
  if (matches(blob, RATE_LIMIT_PATTERNS)) return "rate_limited";
  if (matches(blob, INFRA_PATTERNS)) return "infrastructure";
  if (matches(blob, TOXIC_PATTERNS)) return "toxic_anomaly";
  if (matches(blob, CRED_INVALID_PATTERNS)) return "credential_invalid";
  if (matches(blob, SITE_CHANGE_PATTERNS)) return "site_change";
  return "unknown";
}

/** Return the remediation strategy string for a triage category. */
export function getRemediation(category: TriageCategory): string {
  return REMEDIATION[category] ?? REMEDIATION.unknown;
}

/**
 * Group similar error strings together based on word-overlap (Jaccard similarity).
 */
export function clusterErrors(errors: string[]): Map<string, string[]> {
  const clusters = new Map<string, string[]>();
  const clusterReps = new Map<string, Set<string>>();

  function getWords(text: string): Set<string> {
    return new Set(text.toLowerCase().match(/\w+/g) ?? []);
  }

  for (const error of errors) {
    const words = getWords(error);
    let matchedCluster: string | null = null;

    for (const [clusterRep, repWords] of clusterReps.entries()) {
      const union = new Set([...words, ...repWords]);
      if (union.size === 0) {
        if (words.size === 0) {
          matchedCluster = clusterRep;
          break;
        }
        continue;
      }
      const intersection = new Set([...words].filter((w) => repWords.has(w)));
      const jaccard = intersection.size / union.size;
      if (jaccard > 0.5) {
        matchedCluster = clusterRep;
        break;
      }
    }

    if (matchedCluster !== null) {
      clusters.get(matchedCluster)!.push(error);
    } else {
      clusters.set(error, [error]);
      clusterReps.set(error, words);
    }
  }

  return clusters;
}
